File: projects/tox4j/src/test/java/im/tox/tox4j/core/exceptions/ToxFriendAddExceptionTest.java
package im.tox.tox4j.core.exceptions;

import im.tox.tox4j.ToxCoreTestBase;
import im.tox.tox4j.core.ToxCore;
import im.tox.tox4j.core.ToxCoreConstants;
import org.junit.Before;
import org.junit.Test;
import scala.runtime.BoxedUnit;

import static org.junit.Assert.assertEquals;

public class ToxFriendAddExceptionTest extends ToxCoreTestBase {

  private byte[] validAddress;

  @Before
  public void setUp() throws Exception {
    validAddress = newTox().getAddress();
  }

  @Test(expected = IllegalArgumentException.class)
  public void testInvalidAddress1() throws Exception {
    try (ToxCore tox = newTox()) {
      tox.addFriend(new byte[1], new byte[1]);
    }
  }

  @Test(expected = IllegalArgumentException.class)
  public void testInvalidAddress2() throws Exception {
    try (ToxCore<BoxedUnit> tox = newTox()) {
      tox.addFriend(new byte[ToxCoreConstants.TOX_ADDRESS_SIZE() - 1], new byte[1]);
    }
  }

  @Test(expected = IllegalArgumentException.class)
  public void testInvalidAddress3() throws Exception {
    try (ToxCore<BoxedUnit> tox = newTox()) {
      tox.addFriend(new byte[ToxCoreConstants.TOX_ADDRESS_SIZE() + 1], new byte[1]);
    }
  }

  @Test
  public void testNull1() throws Exception {
    try (ToxCore<BoxedUnit> tox = newTox()) {
      tox.addFriend(null, new byte[1]);
      fail();
    } catch (ToxFriendAddException e) {
      assertEquals(ToxFriendAddException.Code.NULL, e.code());
    }
  }

  @Test
  public void testNull2() throws Exception {
    try (ToxCore<BoxedUnit> tox = newTox()) {
      tox.addFriend(validAddress, null);
      fail();
    } catch (ToxFriendAddException e) {
      assertEquals(ToxFriendAddException.Code.NULL, e.code());
    }
  }

  @Test
  public void testNot_TooLong1() throws Exception {
    try (ToxCore<BoxedUnit> tox = newTox()) {
      tox.addFriend(validAddress, new byte[ToxCoreConstants.MAX_FRIEND_REQUEST_LENGTH() - 1]);
    }
  }

  @Test
  public void testNot_TooLong2() throws Exception {
    try (ToxCore<BoxedUnit> tox = newTox()) {
      tox.addFriend(validAddress, new byte[ToxCoreConstants.MAX_FRIEND_REQUEST_LENGTH()]);
    }
  }

  @Test
  public void testTooLong() throws Exception {
    try (ToxCore<BoxedUnit> tox = newTox()) {
      tox.addFriend(validAddress, new byte[ToxCoreConstants.MAX_FRIEND_REQUEST_LENGTH() + 1]);
      fail();
    } catch (ToxFriendAddException e) {
      assertEquals(ToxFriendAddException.Code.TOO_LONG, e.code());
    }
  }

  @Test
  public void testNoMessage() throws Exception {
    try (ToxCore<BoxedUnit> tox = newTox()) {
      tox.addFriend(validAddress, new byte[0]);
      fail();
    } catch (ToxFriendAddException e) {
      assertEquals(ToxFriendAddException.Code.NO_MESSAGE, e.code());
    }
  }

  @Test
  public void testOwnKey() throws Exception {
    try (ToxCore<BoxedUnit> tox = newTox()) {
      tox.addFriend(tox.getAddress(), new byte[1]);
      fail();
    } catch (ToxFriendAddException e) {
      assertEquals(ToxFriendAddException.Code.OWN_KEY, e.code());
    }
  }

  @Test
  public void testAlreadySent() throws Exception {
    try (ToxCore<BoxedUnit> tox = newTox()) {
      tox.addFriend(validAddress, new byte[1]);
      tox.addFriend(validAddress, new byte[1]);
      fail();
    } catch (ToxFriendAddException e) {
      assertEquals(ToxFriendAddException.Code.ALREADY_SENT, e.code());
    }
  }

  @Test
  public void testBadChecksum() throws Exception {
    try (ToxCore<BoxedUnit> tox = newTox()) {
      validAddress[0]++;
      tox.addFriend(validAddress, new byte[1]);
      fail();
    } catch (ToxFriendAddException e) {
      assertEquals(ToxFriendAddException.Code.BAD_CHECKSUM, e.code());
    }
  }

  @Test
  public void testSetNewNospam() throws Exception {
    try (ToxCore<BoxedUnit> tox = newTox()) {
      ToxCore friend = newTox();
      friend.setNoSpam(12345678);
      tox.addFriend(friend.getAddress(), new byte[1]);
      friend.setNoSpam(87654321);
      tox.addFriend(friend.getAddress(), new byte[1]);
      fail();
    } catch (ToxFriendAddException e) {
      assertEquals(ToxFriendAddException.Code.SET_NEW_NOSPAM, e.code());
    }
  }

  @Test
  public void testMalloc() throws Exception {
    // Can't test this.
    new ToxFriendAddException(ToxFriendAddException.Code.MALLOC).code();
  }

}

File: projects/tox4j/src/test/java/im/tox/tox4j/core/exceptions/ToxFriendAddExceptionTest.scala
package im.tox.tox4j.core.exceptions

import im.tox.tox4j.core.{ToxCoreConstants, ToxCoreFactory}
import im.tox.tox4j.testing.ToxTestMixin
import org.scalatest.FunSuite

final class ToxFriendAddExceptionTest extends FunSuite with ToxTestMixin {
  private val validAddress = ToxCoreFactory.withTox(_.getAddress)

  test("InvalidAddress1") {
    intercept[IllegalArgumentException] {
      ToxCoreFactory.withTox(
        _.addFriend(Array.ofDim[Byte](1), Array.ofDim[Byte](1))
      )
    }
  }

  test("InvalidAddress2") {
    intercept[IllegalArgumentException] {
      ToxCoreFactory.withTox(
        _.addFriend(new Array[Byte](ToxCoreConstants.AddressSize - 1), new Array[Byte](1))
      )
    }
  }

  test("InvalidAddress3") {
    intercept[IllegalArgumentException] {
      ToxCoreFactory.withTox(
        _.addFriend(new Array[Byte](ToxCoreConstants.AddressSize + 1), new Array[Byte](1))
      )
    }
  }

  test("Null1") {
    interceptWithTox(ToxFriendAddException.Code.NULL)(
      _.addFriend(null, new Array[Byte](1))
    )
  }

  test("Null2") {
    interceptWithTox(ToxFriendAddException.Code.NULL)(
      _.addFriend(validAddress, null)
    )
  }

  test("Not_TooLong1") {
    ToxCoreFactory.withTox(
      _.addFriend(validAddress, new Array[Byte](ToxCoreConstants.MaxFriendRequestLength - 1))
    )
  }

  test("Not_TooLong2") {
    ToxCoreFactory.withTox(
      _.addFriend(validAddress, new Array[Byte](ToxCoreConstants.MaxFriendRequestLength))
    )
  }

  test("TooLong") {
    interceptWithTox(ToxFriendAddException.Code.TOO_LONG)(
      _.addFriend(validAddress, new Array[Byte](ToxCoreConstants.MaxFriendRequestLength + 1))
    )
  }

  test("NoMessage") {
    interceptWithTox(ToxFriendAddException.Code.NO_MESSAGE)(
      _.addFriend(validAddress, "".getBytes)
    )
  }

  test("OwnKey") {
    interceptWithTox(ToxFriendAddException.Code.OWN_KEY) { tox =>
      tox.addFriend(tox.getAddress, "hello".getBytes)
    }
  }

  test("AlreadySent") {
    interceptWithTox(ToxFriendAddException.Code.ALREADY_SENT) { tox =>
      tox.addFriend(validAddress, "hello".getBytes)
      tox.addFriend(validAddress, "hello".getBytes)
    }
  }

  test("BadChecksum") {
    interceptWithTox(ToxFriendAddException.Code.BAD_CHECKSUM)(
      _.addFriend(validAddress.updated(0, (validAddress(0) + 1).toByte), "hello".getBytes)
    )
  }

  test("SetNewNospam") {
    interceptWithTox(ToxFriendAddException.Code.SET_NEW_NOSPAM) { tox =>
      ToxCoreFactory.withTox { friend =>
        friend.setNospam(12345678)
        tox.addFriend(friend.getAddress, "hello".getBytes)
        friend.setNospam(87654321)
        tox.addFriend(friend.getAddress, "hello".getBytes)
      }
    }
  }

}
